test/e2e/common/setup.go

```diff
@@ -156,83 +146,129 @@ func setCacertKeys(dir string, data map[string][]byte) error {
+				}
+				return nil
+			}, retry.Timeout(5*time.Minute), retry.Delay(1*time.Second)); err != nil {
+				return "", "", fmt.Errorf("could not update federation-controller (cluster=%s): %v", remoteCluster.Name(), err)
+			}
+		}
+		return gatewayIP, remoteClusterName, nil
+	}
+	return func(ctx resource.Context) error {
+		var g errgroup.Group
+		for idx, localCluster := range ctx.Clusters() {
+			helmUpgradeCmd := exec.Command("helm", "upgrade", "--install", "--wait",
+				"-n", "istio-system",
+				"federation",
+				fmt.Sprintf("%s/chart", rootDir),
+				fmt.Sprintf("--values=%s/test/testdata/federation-controller.yaml", rootDir),
+				"--set", fmt.Sprintf("federation.configMode=%s", configMode),
+				"--set", fmt.Sprintf("image.repository=%s/federation-controller", testHub),
+				"--set", fmt.Sprintf("image.tag=%s", testTag))
+			if configureRemotePeer {
+				gatewayIP, remoteClusterName, err := getRemoteNetworkAndIngressIP(ctx, localCluster)
+				if err != nil {
+					return err
+				}
+				helmUpgradeCmd.Args = append(helmUpgradeCmd.Args,
+					"--set", fmt.Sprintf("federation.meshPeers.remote.addresses[0]=%s", gatewayIP),
+					"--set", fmt.Sprintf("federation.meshPeers.remote.network=%s", remoteClusterName))
 			}
+			helmUpgradeCmd.Env = os.Environ()
+			helmUpgradeCmd.Env = append(helmUpgradeCmd.Env, fmt.Sprintf("KUBECONFIG=%s/test/%s.kubeconfig", rootDir, clusterNames[idx]))
+			g.Go(func() error {
+				if out, err := helmUpgradeCmd.CombinedOutput(); err != nil {
+					return fmt.Errorf("failed to upgrade federation controller (cluster=%s): %v, %v", clusterNames[idx], string(out), err)
+				}
+				return nil
+			})
 		}
-		helmUpgradeCmd := exec.Command("helm", "upgrade", "-n", "istio-system",
-			fmt.Sprintf("%s-federation-controller", clusterNames[idx]),
-			fmt.Sprintf("%s/chart", rootDir),
-			fmt.Sprintf("--values=%s/test/testdata/federation-controller.yaml", rootDir),
-			"--set", fmt.Sprintf("federation.meshPeers.remote.addresses[0]=%s", gatewayIP),
-			"--set", fmt.Sprintf("federation.meshPeers.remote.network=%s", remoteClusterName),
-			"--set", fmt.Sprintf("image.repository=%s/federation-controller", testHub),
-			"--set", fmt.Sprintf("image.tag=%s", testTag))
-		helmUpgradeCmd.Env = os.Environ()
-		helmUpgradeCmd.Env = append(helmUpgradeCmd.Env, fmt.Sprintf("KUBECONFIG=%s/test/%s.kubeconfig", rootDir, clusterNames[idx]))
-		if out, err := helmUpgradeCmd.CombinedOutput(); err != nil {
-			return fmt.Errorf("failed to upgrade federation controller: %v, %v", string(out), err)
+		if err := g.Wait(); err != nil {
+			return err
 		}
+		ctx.Cleanup(func() {
+			for idx := range ctx.Clusters() {
+				helmUninstallCmd := exec.Command("helm", "uninstall", "federation", "-n", "istio-system")
+				helmUninstallCmd.Env = os.Environ()
+				helmUninstallCmd.Env = append(helmUninstallCmd.Env, fmt.Sprintf("KUBECONFIG=%s/test/%s.kubeconfig", rootDir, clusterNames[idx]))
```

Comment:
Does the `helmUninstallCmd` command need to be run too?

Reply:
Yes, of course! It's just a mistake. Thanks.

Reply:
Done